docs/fsharp-cheatsheet.md

```diff
@@ -555,136 +555,158 @@ Another way of implementing interfaces is to use *object expressions*.
-If a new `namespace` is specified, all elements following will be part of the new namespace.
+Namespaces are simply dotted names that prefix `type` and `module` declarations to allow for hierarchical scoping.
+The first `namespace` directives must be placed at the top of the file. Subsequent `namespace` directives either:
+(a) create a sub-namespace (keeping prior declarations in scope); or (b) create a new namespace (and all prior declarations are discarded from scope).
 
     namespace MyNamespace
 
+    module MyModule =  // MyNamspace.MyModule
+        let myLet = ...  // MyNamspace.MyModule.myLet
+
     namespace MyNamespace.SubNamespace
 
-They can also be specified in a file-level [`module`](#CodeOrganization_Modules) definition, but no further `namespace` declarations may follow.
+    namespace MyNewNamespace  // a new namespace
+
+A top-level [`module`](#CodeOrganization_Modules) declaration can bear its namespace as a dotted prefix:
 
     module MyNamespace.SubNamespace.Functions
 
 <a name="CodeOrganization_OpenAndAutoOpen"></a>
 ### Open and AutoOpen
 
-The `open` keyword can be used with `module`, `namespace`, and `type`.
+The `open` keyword can be used on `module`, `namespace`, and `type`.
 
-    open System.Diagnostics  // open namespace
-    let stopwatch = Stopwatch.StartNew()
-    ---
-    module MyModule =
-        type DU1 = A | B | C
-        type DU2 = D | E | F
+    module Groceries =
+        type Fruit =
+            | Apple
+            | Banana
 
-    open type MyModule.DU1
-    let du1 = A
-    let duNotDefined = D  // 'D' not defined
-    open MyModule
-    let du2 = D
+    let fruit1 = Groceries.Banana
+    open Groceries  // module
+    let fruit2 = Apple
+    ---
+    open System.Diagnostics  // namespace
+    let stopwatch = Stopwatch.StartNew()  // Stopwatch is accessible
+    ---
+    open type System.Text.RegularExpressions.Regex  // type
+    let isHttp url = IsMatch("^https?:", url)  // Regex.IsMatch directly accessible
 
-Available to `module` elements, is the `AutoOpen` attribute. This alleviates the need for an `open`; *however* this should be used cautiously,
-as all following declarations will be immediately brought into the global namespace and cause conflicts.
+Available to `module` declarations only, is the `AutoOpen` attribute, which alleviates the need for an `open`.
 
     [<AutoOpen>]
-    module MyModule =
-        type DU = A | B | C
+    module Groceries =
+        type Fruit =
+            | Apple
+            | Banana
 
-    let du = A
+    let fruit = Banana
+
+*However*, `AutoOpen` should be used cautiously. When an `open` or `AutoOpen` is used, all declarations in the containing element
+will be brought into scope. This can lead to [shadowing](https://en.wikipedia.org/wiki/Variable_shadowing); where the last
+named declaration replaces all prior identically-named declarations. There is *no* error - or even a warning - in F#, when shadowing occurs.
```

Comment:
This secontion is pretty good now.... BUT!

maybe shadowing deserves an individual section? you used `open type` above for another reason, but for me a key thing that trips people up is how DU cases shadow others and/or type names. I always used to open a DU's module, but really `open type` on the DU brings less stuff in.

Not sure the "AuotOpen would sidestep thi"s is going to be grokked by someon scanning a cheatsheet...

Reply:
Maybe AutoOpen doesn't need a place on a cheat sheet?

Reply:
That may well be the answer (not mentioning it, and conserving space for better things)

However for me AutoOpen can be an important part of the overall tools in the toolbox, e.g.:
- if you want to split something over 2 files (or 2 projects), but still have them in the same namespace, you can't have two "partial" modules with the same name
- so you use a namespace
- but then you can't have helper functions at top level
- so AutoOpen allows each file to contribute helpers to the top level of the namespace

I'm really not sure what the solution is, and I can't think of a sufficiently succinct example. I guess the main thing is not to give it prominence without conveying that it's way down the list in terms of being the answer to problems...

Reply:
Thanks